### ISSUE TITLE
RFE: add a udev property we check before pulling in systemd-backlight@.service for led/backlight devices.

### ISSUE THREAD

**Issue by tasmya** [systemd/systemd#31110]
### systemd version the issue has been seen with

255 (255.3-1 Arch)

### Used distribution

Arch Linux

### Linux kernel version used

6.7.1-zen1-1-zen

### CPU architectures issue was seen on

x86_64

### Component

systemd

### Expected behaviour you didn't see

After updating the package to the latest version, systemd-backlight slows the startup by around 40 seconds. Reverting to the earlier version 255.2 or masking the service solve the issue.

### Unexpected behaviour you saw

Delay while booting

**Comment by bluca** [comment]
Please provide debug level logs of the service

**Comment by tasmya** [comment]
Instead of a single systemd-backlight@leds:rgb:kbd_backlight.service unit for the whole keyboard (using tuxedo-drivers), a different backlight entry is created for every single key of the keyboard (over 100 entries), which results in a slower startup. Log attached for one of the entries.
[Log.txt](https://github.com/systemd/systemd/files/14075848/Log.txt)

**Comment by bluca** [comment]
probably unintended side effect of https://github.com/systemd/systemd/commit/7ebbaaa7f426b5bb11b4a697898844bf6d8b6ec9 can you try changin the udev rule back manually after installing 255.3 and trying again?

**Comment by tasmya** [comment]
You are right, reverting the udev rule solves the issue

**Comment by poettering** [comment]
Uh, what? your kbd exposes hundreds of individual led devices? that sounds very broken? What's going on?

what is "tuxedo-drivers"? Is that an upstream kernel driver?

Sorry, but this is almost certainly a bug with your driver. It shouldn't create so many separate devices in sysfs. That's slow for many ways, systemd-backlight is just the tip of the iceberg.

**Comment by tasmya** [comment]
> Uh, what? your kbd exposes hundreds of individual led devices? that sounds very broken? What's going on?
> 
> what is "tuxedo-drivers"? Is that an upstream kernel driver?
> 
> Sorry, but this is almost certainly a bug with your driver. It shouldn't create so many separate devices in sysfs. That's slow for many ways, systemd-backlight is just the tip of the iceberg.

It is a driver that controls the keyboard of my laptop, which has RGB led that can have a different color for every key. It was working properly until last systemd update.

**Comment by poettering** [comment]
yeah, but the status quo ante was a bug pretty much: we simply didn't process all led devices.

Sorry, but generating so many led devices simply doesn't really scale. each led device creates uevents, that need to be processed, and various apps will have to process them

This is something the driver needs to address. Sorry. Given the driver is out of tree, this even more is a driver problem.

It might be that the "leds" subsystem is simply not prepared for this kind of thing, i.e. large arrays of leds. It's kinda designed for small numbers of leds, where all this is not an issue.

That all said, I'd be OK with adding some propery we check in the udev rules which can be used to turn off pulling in of systemd-backlight@.service for such devices.

i.e. if SYSTEMD_BACKLIGHT=0 is set for a device we'd not do the thing. Happy to review such a patch.

But other than that: flooding the system with a large number of devices doesn't really work, this needs to be revisited by the folks that write the driver. It might need subsystem changes, dunno, but as this is done by the driver right now is simply not workable.

We had similar cases in the past, where memory slots were exposing ridiculous numbers of sysfs devices and then causing a ridiculous amount of work in userspace by sending out uevents from all. This stuff simply shouldn't be done, and needs to be fixed by kernel.

**Comment by poettering** [comment]
i renamed this into an RFE issue, to add such a property, as sugegsted above. But again, this would be a work-around, not a fix for you issue. The fix must be in your driver.

**Comment by tasmya** [comment]
Thank you, I have reported the driver issue

**Comment by tuxedo-wse** [comment]
FYI: I am aware that the current 126 leds entries are a suboptimal solution, that's why I pushed for a better uapi upstream: https://lore.kernel.org/all/e21a7d87-3059-4a51-af04-1062dac977d2@tuxedocomputers.com/

However it could take some time until actuall code hits the kernel and tuxedo-drivers regarding this.

I was not aware of the performance impact when I submitted https://github.com/systemd/systemd/commit/7ebbaaa7f426b5bb11b4a697898844bf6d8b6ec9 You can already notice it with way less kbd_backlights on some devices

I hope it's ok if I'm using this thread to gather some feedback of possible hotfixes until a better solution is here:
1. Revert https://github.com/systemd/systemd/commit/7ebbaaa7f426b5bb11b4a697898844bf6d8b6ec9 for the time being
2. Limit https://github.com/systemd/systemd/commit/7ebbaaa7f426b5bb11b4a697898844bf6d8b6ec9 to a maximum of lets say 4 entries (i.e. only *kbd_backlight, *kbd_backlight_1, *kbd_backlight_2, and *kbd_backlight_3 are recognized by systemd by rewriting the rule)
3. Patch out per-key control from tuxedo-drivers

**Comment by jwrdegoede** [comment]
@poettering @tuxedo-wse I am not in favor of adding a property to tell systemd to not do its save/restore brightness thing just to work around the current *and temporary* weirdness in tuxedo-drivers. Once we add support for such a property, it can likely never be removed. So lets not  add it at all.

@tuxedo-wse you indicated that you would be willing to patch out per-key control from tuxedo drivers, how about simply renaming the led-class devices to not have "kbd_backlight" in their name , e.g. maybe use "key_backlight" instead ?

I don't know if you have any userspace tools relying on this one LED class device per key API, but if you do I assume that you can adjust that userspace to check for both names.

This way the problem of systemd trying to save/restore all of the keys brightness is avoided without needing to add  a special property for this.

**Comment by tuxedo-wse** [comment]
> @poettering @tuxedo-wse I am not in favor of adding a property to tell systemd to not do its save/restore brightness thing just to work around the current _and temporary_ weirdness in tuxedo-drivers. Once we add support for such a property, it can likely never be removed. So lets not add it at all.
> 
> @tuxedo-wse you indicated that you would be willing to patch out per-key control from tuxedo drivers, how about simply renaming the led-class devices to not have "kbd_backlight" in their name , e.g. maybe use "key_backlight" instead ?
> 
> I don't know if you have any userspace tools relying on this one LED class device per key API, but if you do I assume that you can adjust that userspace to check for both names.
> 
> This way the problem of systemd trying to save/restore all of the keys brightness is avoided without needing to add a special property for this.

This would break at least UPower, Powerdevil and our Tuxedo-Control-Center. Only the last one can easily be changed by me. And it would be an even hackier solution then the current situation.

Taking out the per key support is is least liked method from my side, as it worked somewhat before the commit. And taking away features, even if they just somewhat worked, is never well recieved when not providing an alternative.

**Comment by poettering** [comment]
> @poettering @tuxedo-wse I am not in favor of adding a property to tell systemd to not do its save/restore brightness thing just to work around the current _and temporary_ weirdness in tuxedo-drivers. Once we add support for such a property, it can likely never be removed. So lets not add it at all.

Well we would never set the property, it's just a knob users can easily locally disable this stuff...

> @tuxedo-wse you indicated that you would be willing to patch out per-key control from tuxedo drivers, how about simply renaming the led-class devices to not have "kbd_backlight" in their name , e.g. maybe use "key_backlight" instead ?

That otoh is something I'd call "cheating"... Don't flood udev with hundreds of devices, and hence pointless uevents, it's always going to be a waste of resources...

**Comment by jwrdegoede** [comment]
@tuxedo-wse :
> This would break at least UPower, Powerdevil and our Tuxedo-Control-Center. Only the last one can easily be changed by me. And it would be an even hackier solution than the current situation.

At least upower only supports 1 /sys/class/leds/foo::kbd_backlight entry, so if you create 1 per key, then currently upower only controls 1 key which does not seem really useful. Or am I missing something? 

I don't know what powerdevil is doing.

> Taking out the per key support is is least liked method from my side, as it worked somewhat before the commit. And taking away features, even if they just somewhat worked, is never well recieved when not providing an alternative.

Right I'm not suggesting taking it out though, simply renaming the LED class device names so that they no longer trigger `/lib/udev/rules.d/99-systemd.rules`

OTOH since this is a udev rule anyway I guess that adding filtering by some property to the actual udev rule might not be so bad.

@poettering :

> That otoh is something I'd call "cheating"... Don't flood udev with hundreds of devices, and hence pointless uevents, it's always going to be a waste of resources...

I think we all agree that the current solution with 1 LED class device per key needs to be replaced by something new, the question is how do we quickly solve the issue reported here, while working towards replacing the 1 LED class device per key driver with another solution.
